Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm

```diff
@@ -122,6 +128,62 @@
 #endif
     }
 }
+
+void Download::setPlaceholderURL(NSURL *placeholderURL)
+{
+    if (!placeholderURL)
+        return;
+
+    m_hasReceivedPlaceholderURL = true;
+
+    BOOL usingSecurityScopedURL = [placeholderURL startAccessingSecurityScopedResource];
+
+    NSError *error = nil;
+    NSData *data = [placeholderURL bookmarkDataWithOptions:0 includingResourceValuesForKeys:nil relativeToURL:nil error:&error];
+    if (error)
+        RELEASE_LOG_ERROR(Network, "Failed to create bookmark data: %{public}@", error);
+
+    if (usingSecurityScopedURL)
+        [placeholderURL stopAccessingSecurityScopedResource];
+
+    send(Messages::DownloadProxy::DidReceivePlaceholderURL(placeholderURL, span(data)));
+}
+
+void Download::setFinalURL(NSURL *finalURL)
+{
+    if (!finalURL)
+        return;
+
+    m_hasReceivedFinalURL = true;
+
+    BOOL usingSecurityScopedURL = [finalURL startAccessingSecurityScopedResource];
+
+    NSError *error = nil;
+    NSData *data = [finalURL bookmarkDataWithOptions:0 includingResourceValuesForKeys:nil relativeToURL:nil error:&error];
```

Comment:
Does this need to be called in the network process and passed separately?

Reply:
Yes, we send bookmark data for the URL, in case the client wants to be able to locate the file after it has been moved or deleted.